test/file.test.js

```diff
@@ -258,6 +258,293 @@ describe('file', function() {
+                await fs.unlink(path.resolve('./test-transfer-file-status-200-truncated.dat'));
+            } catch (e) {
+                console.log(e);
+            }
+        });
+        it('status-200-stream-retry', async function() {
+            nock('http://test-status-200-stream-retry')
+                .head('/path/to/file.ext')
+                .reply(200, "OK",{
+                    'content-type': 'image/jpeg',
+                    'content-length': 11,
+                    'content-disposition': 'attachment; filename="image-file-1.jpg"',
+                    'last-modified': 'Wed, 21 Oct 2015 07:28:00 GMT',
+                    'etag': ''
+                });
+            nock('http://test-status-200-stream-retry')
+                .get('/path/to/file.ext')
+                .twice()
+                .reply(200, 'hello world', {
+                    'content-type': 'image/jpeg',
+                    'content-length': 11
+                });
+
+            testSetResponseBodyOverride("GET", createErrorReadable(Error('read error')));
+            await downloadFileConcurrently('http://test-status-200-stream-retry/path/to/file.ext', path.resolve('./test-transfer-file-filestream.dat'));
+            const result = await fs.readFile(path.resolve('./test-transfer-file-filestream.dat'), 'utf8');
+            assert.strictEqual(result, 'hello world');
+
+            try {
+                await fs.unlink(path.resolve('./test-transfer-file-filestream.dat'));
+            } catch (e) {
+                console.log(e);
+            }
+        });
+        it('status-404', async function() {
+            try {
+                nock('http://test-status-404')
+                    .head('/path/to/file.ext')
+                    .reply(200, "OK",{
+                        'content-type': 'image/jpeg',
+                        'content-length': 11,
+                        'content-disposition': 'attachment; filename="image-file-1.jpg"',
+                        'last-modified': 'Wed, 21 Oct 2015 07:28:00 GMT',
+                        'etag': ''
+                    });
+                nock('http://test-status-404')
+                    .get('/path/to/file.ext')
+                    .reply(404, 'hello world');
+
+                await downloadFileConcurrently('http://test-status-404/path/to/file.ext', path.resolve('./test-transfer-file-status-404.dat'));
+            } catch (e) {
+                assert.ok(e.message.includes('GET'), e.message);
+                assert.ok(e.message.includes('failed with status 404'));
+                const result = await fs.readFile(path.resolve('./test-transfer-file-status-404.dat'), 'utf8');
+                assert.strictEqual(result, '');
+            }
+
+            try {
+                await fs.unlink(path.resolve('./test-transfer-file-status-404.dat'));
+            } catch (e) {
+                console.log(e);
+            }
+        });
+        it('status-404-retry', async function() {
```

Comment:
I think the rest I didn't get to fix yet. there are two main changes that impact tests that I have identified so far:

1. downloadFileConcurrently expects server to respond with a content-length header
        - impact: have to adjust nocks in tests
2. downloadFileConcurrently does a head request for the file size (BlockDownload requires file size for determining parts, etc.)
        - impact: have to adjust nocks in tests